core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -932,7 +932,10 @@ public ClusterState execute(ClusterState currentState) {
                     if (modified) {
                         stateBuilder.nodes(nodesBuilder);
                     }
-                    return stateBuilder.build();
+                    currentState = stateBuilder.build();
```

Comment:
We do need to publish a cluster state to seal the mastership on the joining nodes. I will add a big fat comment about this here after this is committed (to not be in your way when you merge). What I meant is avoiding the reroute, which was accumulated before for all the joins. But ok to leave as is for now if you have concerns.